File: Android/app/src/main/java/com/hhp227/datemate/data/UserRemoteDataSource.kt
package com.hhp227.datemate.data

import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.hhp227.datemate.common.Resource
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.tasks.await

class UserRemoteDataSource private constructor(
    private val firebaseAuth: FirebaseAuth
) {
    val userStateFlow: Flow<FirebaseUser?> = callbackFlow {
        val listener = FirebaseAuth.AuthStateListener { firebaseAuth ->
            trySend(firebaseAuth.currentUser)
        }

        firebaseAuth.addAuthStateListener(listener)
        awaitClose { firebaseAuth.removeAuthStateListener(listener) }
    }

    fun signIn(email: String, password: String): Flow<Resource<FirebaseUser>> = flow {
        try {
            val result = firebaseAuth
                .signInWithEmailAndPassword(email, password)
                .await()
            val user = result.user

            if (user != null) {
                emit(Resource.Success(user))
            } else {
                emit(Resource.Error("로그인 실패"))
            }
        } catch (e: Exception) {
            emit(Resource.Error("로그인 오류: ${e.message}"))
        }
    }
        .onStart { emit(Resource.Loading()) }

    fun signOut(): Flow<Resource<Boolean>> = flow {
        try {
            firebaseAuth.signOut()
            emit(Resource.Success(true))
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: ""))
        }
    }
        .onStart { emit(Resource.Loading()) }

    suspend fun signUp(email: String, password: String): Task<AuthResult> {
        return firebaseAuth.createUserWithEmailAndPassword(email, password)
    }

    companion object {
        @Volatile
        private var instance: UserRemoteDataSource? = null
        fun getInstance(auth: FirebaseAuth) =
            instance ?: synchronized(this) { instance ?: UserRemoteDataSource(auth).also { instance = it } }
    }
}

File: iOS/DateMate/Data/UserRemoteDataSource.swift
//
//  UserRemoteDataSource.swift
//  DateMate
//
//  Created by 홍희표 on 2025/11/14.
//

import Foundation
import FirebaseAuth

class UserRemoteDataSource {
    private let firebaseAuth: Auth
    
    func signIn(email: String, password: String) async throws -> AuthDataResult {
        return try await withCheckedThrowingContinuation { continuation in
            firebaseAuth.signIn(withEmail: email, password: password) { authResult, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let authResult = authResult {
                    continuation.resume(returning: authResult)
                }
            }
        }
    }
    
    func signUp(email: String, password: String) async throws -> AuthDataResult {
        return try await withCheckedThrowingContinuation { continuation in
            firebaseAuth.createUser(withEmail: email, password: password) { authResult, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let authResult = authResult {
                    continuation.resume(returning: authResult)
                }
            }
        }
    }

    func addAuthStateListener(_ listener: @escaping (Auth?, User?) -> Void) -> AuthStateDidChangeListenerHandle {
        firebaseAuth.addStateDidChangeListener(listener)
    }

    func removeAuthStateListener(_ handle: AuthStateDidChangeListenerHandle) {
        firebaseAuth.removeStateDidChangeListener(handle)
    }
    
    init(_ firebaseAuth: Auth) {
        self.firebaseAuth = firebaseAuth
    }
    
    private static var instance: UserRemoteDataSource? = nil
    
    static func getInstance(auth: Auth) -> UserRemoteDataSource {
        if let instance = self.instance {
            return instance
        } else {
            let userDataSource = UserRemoteDataSource(auth)
            self.instance = userDataSource
            return userDataSource
        }
    }
}
